test/unittest/tests/wrappers/mock_session.h

```diff
@@ -10,9 +10,26 @@
 #define WT_MOCK_SESSION_H
 
 #include <memory>
+#include <string>
+#include <vector>
+
 #include "wt_internal.h"
 #include "mock_connection.h"
 
+int handleWiredTigerError(
+  WT_EVENT_HANDLER *handler, WT_SESSION *session, int error, const char *message);
+int handleWiredTigerMessage(WT_EVENT_HANDLER *handler, WT_SESSION *session, const char *message);
+
+/* Forward declare the class so we can include it in our wrapper. */
+class MockSession;
+
+/* This is a convenience type that lets us get back to our mock from the handler callback. */
+struct event_handler_wrapper {
+    WT_EVENT_HANDLER handler;
+    MockSession *mock_session;
+};
+typedef struct event_handler_wrapper handler_wrap;
```

Comment:
Do we need to get back to the `mock_session` from the wrapper, if session owns event handlers anyways?

My initial thought was to just directly initialise a WT_EVENT_HANDLER in the session as a new argument rather than the wrapper.

Reply:
In C++ (unlike C), typedef struct is not required, so we can remove the line 

`typedef struct event_handler_wrapper handler_wrap;`

and simply use `event_handler_wrapper` instead of `handler_wrap` in each case

Reply:
The `event_handler_wrapper` struct is required as we need to be able to cast the `WT_EVENT_HANDLER*` callback parameter to a `struct` (`event_handler_wrapper`) that contains a pointer to the mock session. The downcast only works because we have a struct with a first member which is an `WT_EVENT_HANDLER`.

Reply:
Yeah this logic was the most tricky to get to work, obviously the struct wrapper was the best option but I tried two other options:

- A static message store that would be accessed any time anyone called the handler, given unit tests are singly threaded this seemed okay but it also ended up being a pain due to compilation units.
- Add the callback functions to the MockSession itself. This would've been the cleanest but turns out it breaks C++ and is undefined behaviour, technically doable with some horrific casting.

Anyway because the WT_SESSION_IMPL doesn't have a place to save the messages that isn't hacky even if you set the handler on the session you'd still not be able to access the last messages that came through at the unit test level.

Reply:
Fixed the typedef struct in: 281164070